ecs_client_test.go

```diff
@@ -39,49 +64,220 @@ func TestECSClientRegisterAndDeregisterTaskDefinition(t *testing.T) {
+
+			defer c.Close(tctx)
+
+			tCase(tctx, t, c)
 		})
+	}
+
+	registerIn := &ecs.RegisterTaskDefinitionInput{
+		ContainerDefinitions: []*ecs.ContainerDefinition{
+			{
+				Command: []*string{aws.String("echo"), aws.String("foo")},
+				Image:   aws.String("busybox"),
+				Name:    aws.String("print_foo"),
+			},
+		},
+		Cpu:    aws.String("128"),
+		Memory: aws.String("4"),
+		Family: aws.String(makeFamily()),
+	}
+
+	registerOut, err := c.RegisterTaskDefinition(ctx, registerIn)
+	require.NoError(t, err)
+	require.NotZero(t, registerOut)
+	require.NotZero(t, registerOut.TaskDefinition)
+
+	defer func() {
+		cleanupTaskDefinition(ctx, t, c, registerOut)
+		err := c.Close(ctx)
 		require.NoError(t, err)
-		require.NotZero(t, out)
-
-		defer func() {
-			if out != nil && out.TaskDefinition != nil && out.TaskDefinition.TaskDefinitionArn != nil {
-				out, err := c.DeregisterTaskDefinition(ctx, &ecs.DeregisterTaskDefinitionInput{
-					TaskDefinition: out.TaskDefinition.TaskDefinitionArn,
-				})
-				require.NoError(t, err)
-				require.NotZero(t, out)
-			}
-		}()
-
-		require.NotZero(t, out.TaskDefinition)
-		require.NotZero(t, out.TaskDefinition.Status)
-		assert.Equal(t, "ACTIVE", *out.TaskDefinition.Status)
-		require.NotZero(t, out.TaskDefinition.RegisteredAt)
-		assert.NotZero(t, *out.TaskDefinition.RegisteredAt)
-	})
+	}()
+
+	for tName, tCase := range map[string]func(context.Context, *testing.T, *BasicECSClient){
+		"RunTaskFailsWithInvalidInput": func(ctx context.Context, t *testing.T, c *BasicECSClient) {
+			out, err := c.RunTask(ctx, &ecs.RunTaskInput{})
+			assert.Error(t, err)
+			assert.Zero(t, out)
+		},
+		"DescribeTasksFailsWithInvalidInput": func(ctx context.Context, t *testing.T, c *BasicECSClient) {
+			out, err := c.DescribeTasks(ctx, &ecs.DescribeTasksInput{})
+			assert.Error(t, err)
+			assert.Zero(t, out)
+		},
+		"StopTaskFailsWithInvalidInput": func(ctx context.Context, t *testing.T, c *BasicECSClient) {
+			out, err := c.StopTask(ctx, &ecs.StopTaskInput{})
+			assert.Error(t, err)
+			assert.Zero(t, out)
+		},
+		"RunTaskFailsWithValidButNonexistentInput": func(ctx context.Context, t *testing.T, c *BasicECSClient) {
+			out, err := c.RunTask(ctx, &ecs.RunTaskInput{
+				Cluster:        aws.String(os.Getenv("AWS_ECS_CLUSTER")),
+				TaskDefinition: aws.String(utility.RandomString()),
+			})
+			assert.Error(t, err)
+			assert.Zero(t, out)
+		},
+		"RunTaskFailsWithoutCluster": func(ctx context.Context, t *testing.T, c *BasicECSClient) {
+			require.NotZero(t, registerOut.TaskDefinition.Status)
+			assert.Equal(t, "ACTIVE", *registerOut.TaskDefinition.Status)
+
+			runOut, err := c.RunTask(ctx, &ecs.RunTaskInput{
+				TaskDefinition: registerOut.TaskDefinition.TaskDefinitionArn,
+			})
+
+			require.Error(t, err)
+			require.Zero(t, runOut)
+
+			defer cleanupTask(ctx, t, c, runOut)
+		},
+		"DescribeTasksFailsWithNoClusterAndValidButNonexistentInput": func(ctx context.Context, t *testing.T, c *BasicECSClient) {
```

Comment:
I don't think we actually have to test this case. This is testing the cluster-defaulting behavior, but that's not particularly important to us since we'll always specify the cluster name rather than use whatever the default cluster is.

Reply:
Okay, sounds good.